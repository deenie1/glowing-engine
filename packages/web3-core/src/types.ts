import { Socket } from 'net';
import {
	Web3APISpec,
	JsonRpcPayload,
	JsonRpcResponse,
	JsonRpcResult,
	Web3BaseProvider,
} from 'web3-common';

export type LegacyRequestProvider = {
	request: <R = JsonRpcResult, P = unknown>(
		payload: JsonRpcPayload<P>,
		cb: (err: Error | null, response: JsonRpcResponse<R>) => void,
	) => void;
};

export type LegacySendProvider = {
	send: <R = JsonRpcResult, P = unknown>(
		payload: JsonRpcPayload<P>,
		cb: (err: Error | null, response: JsonRpcResponse<R>) => void,
	) => void;
};

export type LegacySendAsyncProvider = {
	sendAsync: <R = JsonRpcResult, P = unknown>(
		payload: JsonRpcPayload<P>,
	) => Promise<JsonRpcResponse<R>>;
};

export type SupportedProviders<API extends Web3APISpec> =
	| Web3BaseProvider<API>
	| LegacyRequestProvider
	| LegacySendProvider
	| LegacySendAsyncProvider;

export type Web3BaseProviderConstructor = new <API extends Web3APISpec>(
	url: string,
	net?: Socket,
) => Web3BaseProvider<API>;
