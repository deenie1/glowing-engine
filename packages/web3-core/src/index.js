/*
    This file is part of web3.js.

    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
 * @file index.js
 * @author Fabian Vogelsteller <fabian@ethereum.org>
 * @date 2017
 */

const { Manager, BatchManager } = require("web3-core-requestmanager");
const extend = require("./extend");






class Core {
  constructor (...args) {
    // sets _requestmanager etc
    if (args[0] && args[0]._requestManager) {
        this._requestManager = args[0]._requestManager;
    } else {
        this._requestManager = new Manager(args[0], args[1]);
    }

    // add givenProvider
    this.givenProvider = Manager.givenProvider;
    this.providers = Manager.providers;

    this._provider = this._requestManager.provider;

    // attach batch request creation
    this.BatchRequest = BatchManager.bind(null, this._requestManager);

    // attach extend function
    this.extend = extend(this);
  }

  setRequestManager (manager) {
    this._requestManager = manager;
    this._provider = manager.provider;
  }

  set curentProvider (newProvider) {
    return this.setProvider(newProvider);
  }

  get curentProvider () {
    return this._provider;
  }

  static get providers () {
    return Manager.providers
  }

  static get givenProvider () {
    return Manager.givenProvider
  }

  setProvider (provider, net) {
    this._requestManager.setProvider(provider, net);
    this._provider = this._requestManager.provider;
    this.extend = extend(this);
    return true;
  }

  addProviders () {
    this.givenProvider = Manager.givenProvider;
    this.providers = Manager.providers;
  }

  static addProviders (pkg) {
    pkg.givenProvider = Manager.givenProvider;
    pkg.providers = Manager.providers;
  }

  static packageInit (pkg, args) {
    args = Array.prototype.slice.call(args);

    if (!pkg) {
        throw new Error('You need to instantiate using the "new" keyword.');
    }

    // make property of pkg._provider, which can properly set providers
    Object.defineProperty(pkg, 'currentProvider', {
        get: () => {
            return pkg._provider;
        },
        set: (value) => {
            return pkg.setProvider(value);
        },
        enumerable: true,
        configurable: true
    });

    // inherit from parent package or create a new RequestManager
    if (args[0] && args[0]._requestManager) {
        pkg._requestManager = args[0]._requestManager;
    } else {
        pkg._requestManager = new Manager(args[0], args[1]);
    }

    // add givenProvider
    pkg.givenProvider = Manager.givenProvider;
    pkg.providers = Manager.providers;

    pkg._provider = pkg._requestManager.provider;

    // add SETPROVIDER function (don't overwrite if already existing)
    if (!pkg.setProvider) {
        pkg.setProvider = (provider, net) => {
            pkg._requestManager.setProvider(provider, net);
            pkg._provider = pkg._requestManager.provider;
            return true;
        };
    }

    pkg.setRequestManager = (manager) => {
        pkg._requestManager = manager;
        pkg._provider = manager.provider;
    };

    // attach batch request creation
    pkg.BatchRequest = BatchManager.bind(null, pkg._requestManager);

    // attach extend function
    pkg.extend = extend(pkg);
  }



}

module.exports = Core
