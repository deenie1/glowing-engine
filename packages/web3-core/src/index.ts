/*
This file is part of web3.js.

web3.js is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

web3.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/

export * from './web3_config';
export * from './web3_request_manager';
export * from './web3_subscription_manager';
export * from './web3_subscriptions';
export * from './web3_context';
export * from './web3_batch_request';
export * from './utils';
export * from './types';
export * from './formatters';
export * from './web3_promi_event';
export * from './web3_event_emitter';

// For backward usability export as namespace
export * as formatters from './formatters';
