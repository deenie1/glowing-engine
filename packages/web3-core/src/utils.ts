import { Web3APISpec, Web3BaseProvider } from 'web3-common';
import {
	LegacyRequestProvider,
	LegacySendAsyncProvider,
	LegacySendProvider,
	SupportedProviders,
} from './types';

export const isWeb3Provider = <API extends Web3APISpec>(
	provider: SupportedProviders<API>,
): provider is Web3BaseProvider<API> => Web3BaseProvider.isWeb3Provider(provider);

export const isLegacyRequestProvider = <API extends Web3APISpec>(
	provider: SupportedProviders<API>,
): provider is LegacyRequestProvider => 'request' in provider;

export const isLegacySendProvider = <API extends Web3APISpec>(
	provider: SupportedProviders<API>,
): provider is LegacySendProvider => 'send' in provider;

export const isLegacySendAsyncProvider = <API extends Web3APISpec>(
	provider: SupportedProviders<API>,
): provider is LegacySendAsyncProvider => 'sendAsync' in provider;

export const isSupportSubscriptions = <API extends Web3APISpec>(
	provider: SupportedProviders<API>,
): boolean => {
	if (isWeb3Provider<API>(provider)) {
		return provider.supportsSubscriptions();
	}

	if ('on' in provider) {
		return true;
	}

	return false;
};
