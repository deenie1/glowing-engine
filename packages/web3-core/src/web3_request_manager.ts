/*
This file is part of web3.js.

web3.js is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

web3.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Socket } from 'net';
import {
	EthExecutionAPI,
	InvalidResponseError,
	jsonRpc,
	JsonRpcBatchRequest,
	JsonRpcBatchResponse,
	JsonRpcPayload,
	JsonRpcResponse,
	ProviderError,
	ResponseError,
	Web3APIMethod,
	Web3APIPayload,
	Web3APIRequest,
	Web3APIReturnType,
	Web3APISpec,
	Web3BaseProvider,
	Web3EventEmitter,
} from 'web3-common';
import HttpProvider from 'web3-providers-http';
import WSProvider from 'web3-providers-ws';
import IpcProvider from 'web3-providers-ipc';
import { SupportedProviders, Web3BaseProviderConstructor } from './types';
import {
	isLegacyRequestProvider,
	isLegacySendAsyncProvider,
	isLegacySendProvider,
	isWeb3Provider,
} from './utils';

export enum Web3RequestManagerEvent {
	PROVIDER_CHANGED = 'PROVIDER_CHANGED',
	BEFORE_PROVIDER_CHANGE = 'BEFORE_PROVIDER_CHANGE',
}

const availableProviders = {
	HttpProvider: HttpProvider as Web3BaseProviderConstructor,
	WebsocketProvider: WSProvider as Web3BaseProviderConstructor,
	IpcProvider: IpcProvider as Web3BaseProviderConstructor,
};

export class Web3RequestManager<
	API extends Web3APISpec = EthExecutionAPI,
> extends Web3EventEmitter<{
	[key in Web3RequestManagerEvent]: SupportedProviders<API>;
}> {
	private _provider!: SupportedProviders<API>;

	public constructor(provider?: SupportedProviders<API>, net?: Socket) {
		super();

		if (provider) {
			this.setProvider(provider, net);
		}
	}

	public static get providers() {
		return availableProviders;
	}

	public get provider() {
		if (!this._provider) {
			throw new ProviderError('Provider not available');
		}

		return this._provider;
	}

	// eslint-disable-next-line class-methods-use-this
	public get providers() {
		return availableProviders;
	}

	public setProvider(provider: SupportedProviders<API>, net?: Socket) {
		let newProvider!: Web3BaseProvider<API>;

		// autodetect provider
		if (provider && typeof provider === 'string' && this.providers) {
			// HTTP
			if (/^http(s)?:\/\//i.test(provider)) {
				newProvider = new this.providers.HttpProvider<API>(provider);

				// WS
			} else if (/^ws(s)?:\/\//i.test(provider)) {
				newProvider = new this.providers.WebsocketProvider<API>(provider);

				// IPC
			} else if (typeof net === 'object' && typeof net.connect === 'function') {
				newProvider = new this.providers.IpcProvider<API>(provider, net);
			} else {
				throw new ProviderError(`Can't autodetect provider for "${provider}'"`);
			}
		}

		this.emit(Web3RequestManagerEvent.BEFORE_PROVIDER_CHANGE, this._provider);
		this._provider = newProvider ?? provider ?? null;
		this.emit(Web3RequestManagerEvent.PROVIDER_CHANGED, this._provider);
	}

	public async send<
		Method extends Web3APIMethod<API>,
		ResponseType = Web3APIReturnType<API, Method>,
	>(request: Web3APIRequest<API, Method>): Promise<ResponseType> {
		const response = await this._sendRequest<Method, ResponseType>(request);

		if (jsonRpc.isResponseWithResult(response)) {
			return response.result;
		}

		throw new ResponseError(response);
	}

	public async sendBatch(request: JsonRpcBatchRequest): Promise<JsonRpcBatchResponse<unknown>> {
		const response = await this._sendRequest<never, never>(request);

		return response as JsonRpcBatchResponse<unknown>;
	}

	private async _sendRequest<
		Method extends Web3APIMethod<API>,
		ResponseType = Web3APIReturnType<API, Method>,
	>(
		request: Web3APIRequest<API, Method> | JsonRpcBatchRequest,
	): Promise<JsonRpcResponse<ResponseType>> {
		const { provider } = this;

		const payload = jsonRpc.isBatchRequest(request)
			? jsonRpc.toBatchPayload(request)
			: jsonRpc.toPayload(request);

		if (isWeb3Provider(provider)) {
			const response = await provider.request<Method, ResponseType>(
				payload as Web3APIPayload<API, Method>,
			);

			return this._processJsonRpcResponse(payload, response);
		}

		// TODO: This should be deprecated and removed.
		if (isLegacyRequestProvider(provider)) {
			return new Promise<JsonRpcResponse<ResponseType>>((resolve): void => {
				provider.request<ResponseType>(payload, (err, response) => {
					if (err) {
						throw err;
					}

					return resolve(this._processJsonRpcResponse(payload, response));
				});
			});
		}

		// TODO: This should be deprecated and removed.
		if (isLegacySendProvider(provider)) {
			return new Promise<JsonRpcResponse<ResponseType>>((resolve): void => {
				provider.send<ResponseType>(payload, (err, response) => {
					if (err) {
						throw err;
					}

					return resolve(this._processJsonRpcResponse(payload, response));
				});
			});
		}

		// TODO: This should be deprecated and removed.
		if (isLegacySendAsyncProvider(provider)) {
			return provider
				.sendAsync<ResponseType>(payload)
				.then(response => this._processJsonRpcResponse(payload, response));
		}

		throw new ProviderError('Provider does not have a request or send method to use.');
	}

	// eslint-disable-next-line class-methods-use-this
	private _processJsonRpcResponse<ResultType, ErrorType, RequestType>(
		payload: JsonRpcPayload<RequestType>,
		response: JsonRpcResponse<ResultType, ErrorType>,
	): JsonRpcResponse<ResultType> | never {
		if (jsonRpc.isBatchRequest(payload) && !Array.isArray(response)) {
			throw new ResponseError(response, 'Got normal response for a batch request.');
		}

		if (!jsonRpc.isBatchRequest(payload) && Array.isArray(response)) {
			throw new ResponseError(response, 'Got batch response for a normal request.');
		}

		if (jsonRpc.isBatchResponse(response)) {
			return response as JsonRpcBatchResponse<ResultType>;
		}

		if (
			(jsonRpc.isResponseWithError(response) || jsonRpc.isResponseWithResult(response)) &&
			!jsonRpc.isBatchRequest(payload)
		) {
			if (response.id && payload.id !== response.id) {
				throw new InvalidResponseError<ErrorType>(response);
			}
		}

		if (jsonRpc.isResponseWithError<ErrorType>(response)) {
			throw new InvalidResponseError<ErrorType>(response);
		}

		if (jsonRpc.isResponseWithResult<ResultType>(response)) {
			return response;
		}

		throw new ResponseError(response, 'Invalid response');
	}
}
