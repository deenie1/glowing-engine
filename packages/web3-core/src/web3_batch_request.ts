import {
	JsonRpcBatchResponse,
	JsonRpcOptionalRequest,
	DeferredPromise,
	jsonRpc,
	JsonRpcRequest,
	ResponseError,
	OperationAbortError,
	OperationTimeoutError,
} from 'web3-common';
import { Web3RequestManager } from './web3_request_manager';

export const DEFAULT_BATCH_REQUEST_TIMEOUT = 1000;

export class Web3BatchRequest {
	private readonly _requestManager: Web3RequestManager;
	private readonly _requests: Map<
		number,
		{ payload: JsonRpcRequest; promise: DeferredPromise<unknown> }
	>;

	public constructor(requestManager: Web3RequestManager) {
		this._requestManager = requestManager;
		this._requests = new Map();
	}

	public get requests() {
		return [...this._requests.values()].map(r => r.payload);
	}

	public add<ResponseType = unknown>(request: JsonRpcOptionalRequest<unknown>) {
		const payload = jsonRpc.toPayload(request) as JsonRpcRequest;
		const promise = new DeferredPromise<ResponseType>();

		this._requests.set(payload.id as number, { payload, promise });

		return promise;
	}

	// eslint-disable-next-line class-methods-use-this
	public async execute(): Promise<JsonRpcBatchResponse<unknown, unknown>> {
		if (this.requests.length === 0) {
			return Promise.resolve([]);
		}

		const request = new DeferredPromise<JsonRpcBatchResponse<unknown, unknown>>({
			timeout: DEFAULT_BATCH_REQUEST_TIMEOUT,
			eagerStart: true,
			timeoutMessage: 'Batch request timeout',
		});

		this._processBatchRequest(request).catch(err => request.reject(err));

		request.catch((err: Error) => {
			if (err instanceof OperationTimeoutError) {
				this._abortAllRequests('Batch request timeout');
			}

			request.reject(err);
		});

		return request;
	}

	private async _processBatchRequest(
		promise: DeferredPromise<JsonRpcBatchResponse<unknown, unknown>>,
	) {
		const response = await this._requestManager.sendBatch(
			[...this._requests.values()].map(r => r.payload),
		);

		if (response.length !== this._requests.size) {
			this._abortAllRequests('Invalid batch response');

			throw new ResponseError(
				response,
				`Batch request size mismatch the results size. Requests: ${this._requests.size}, Responses: ${response.length}`,
			);
		}

		const requestIds = this.requests
			.map(r => r.id)
			.map(Number)
			.sort((a, b) => a - b);

		const responseIds = response
			.map(r => r.id)
			.map(Number)
			.sort((a, b) => a - b);

		if (JSON.stringify(requestIds) !== JSON.stringify(responseIds)) {
			this._abortAllRequests('Invalid batch response');

			throw new ResponseError(
				response,
				`Batch request mismatch the results. Requests: [${requestIds.join()}], Responses: [${responseIds.join()}]`,
			);
		}

		for (const res of response) {
			if (jsonRpc.isResponseWithResult(res)) {
				this._requests.get(res.id as number)?.promise.resolve(res.result);
			} else if (jsonRpc.isResponseWithError(res)) {
				this._requests.get(res.id as number)?.promise.reject(res.error);
			}
		}

		promise.resolve(response);
	}

	private _abortAllRequests(msg: string) {
		for (const { promise } of this._requests.values()) {
			promise.reject(new OperationAbortError(msg));
		}
	}
}
