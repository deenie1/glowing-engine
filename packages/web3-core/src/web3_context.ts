/*
This file is part of web3.js.

web3.js is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

web3.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/

import {
	Web3APISpec,
	Web3BaseWallet,
	Web3BaseWalletAccount,
	Web3AccountProvider,
} from 'web3-common';
import { HexString, isNullish } from 'web3-utils';
import { SupportedProviders } from './types';
import { isSupportedProvider } from './utils';
// eslint-disable-next-line import/no-cycle
import { Web3Config, Web3ConfigEvent, Web3ConfigOptions } from './web3_config';
import { Web3RequestManager } from './web3_request_manager';
import { Web3SubscriptionConstructor } from './web3_subscriptions';
import { Web3SubscriptionManager } from './web3_subscription_manager';
import { Web3BatchRequest } from './web3_batch_request';

// To avoid circular dependencies, we need to export type from here.
export type Web3ContextObject<
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	API extends Web3APISpec = any,
	RegisteredSubs extends {
		[key: string]: Web3SubscriptionConstructor<API>;
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
	} = any,
> = {
	config: Web3ConfigOptions;
	provider?: SupportedProviders<API> | string;
	requestManager: Web3RequestManager<API>;
	subscriptionManager?: Web3SubscriptionManager<API, RegisteredSubs> | undefined;
	registeredSubscriptions?: RegisteredSubs;
	providers: typeof Web3RequestManager.providers;
	accountProvider?: Web3AccountProvider<Web3BaseWalletAccount>;
	wallet?: Web3BaseWallet<Web3BaseWalletAccount>;
};

export type Web3ContextInitOptions<
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	API extends Web3APISpec = any,
	RegisteredSubs extends {
		[key: string]: Web3SubscriptionConstructor<API>;
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
	} = any,
> = {
	config?: Partial<Web3ConfigOptions>;
	provider?: SupportedProviders<API> | string;
	requestManager?: Web3RequestManager<API>;
	subscriptionManager?: Web3SubscriptionManager<API, RegisteredSubs> | undefined;
	registeredSubscriptions?: RegisteredSubs;
	accountProvider?: Web3AccountProvider<Web3BaseWalletAccount>;
	wallet?: Web3BaseWallet<Web3BaseWalletAccount>;
};

export type Web3ContextConstructor<
	// eslint-disable-next-line no-use-before-define, @typescript-eslint/no-explicit-any
	T extends Web3Context<any>,
	T2 extends unknown[],
> = new (...args: [...extras: T2, context: Web3ContextObject]) => T;

// To avoid circular dependencies, we need to export type from here.
export type Web3ContextFactory<
	// eslint-disable-next-line no-use-before-define, @typescript-eslint/no-explicit-any
	T extends Web3Context<any>,
	T2 extends unknown[],
> = Web3ContextConstructor<T, T2> & {
	fromContextObject(this: Web3ContextConstructor<T, T2>, contextObject: Web3ContextObject): T;
};

export class Web3Context<
	API extends Web3APISpec,
	RegisteredSubs extends {
		[key: string]: Web3SubscriptionConstructor<API>;
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
	} = any,
> extends Web3Config {
	public static readonly providers = Web3RequestManager.providers;
	public static givenProvider?: SupportedProviders<never>;
	public readonly providers = Web3RequestManager.providers;
	private _requestManager: Web3RequestManager<API>;
	private _subscriptionManager?: Web3SubscriptionManager<API, RegisteredSubs>;
	private _accountProvider?: Web3AccountProvider<Web3BaseWalletAccount>;
	private _wallet?: Web3BaseWallet<Web3BaseWalletAccount>;

	public constructor(
		providerOrContext?:
			| string
			| SupportedProviders<API>
			| Web3ContextInitOptions<API, RegisteredSubs>,
	) {
		super();

		// If "providerOrContext" is provided as "string" or an objects matching "SupportedProviders" interface
		if (
			isNullish(providerOrContext) ||
			(typeof providerOrContext === 'string' && providerOrContext.trim() !== '') ||
			isSupportedProvider(providerOrContext as SupportedProviders<API>)
		) {
			this._requestManager = new Web3RequestManager<API>(
				providerOrContext as undefined | string | SupportedProviders<API>,
			);
			this._subscriptionManager = new Web3SubscriptionManager(
				this._requestManager,
				{} as RegisteredSubs,
			);

			return;
		}

		const {
			config,
			provider,
			requestManager,
			subscriptionManager,
			registeredSubscriptions,
			accountProvider,
			wallet,
		} = providerOrContext as Web3ContextInitOptions<API, RegisteredSubs>;

		this.setConfig(config ?? {});

		this._requestManager = requestManager ?? new Web3RequestManager<API>(provider);

		if (subscriptionManager) {
			this._subscriptionManager = subscriptionManager;
		} else if (registeredSubscriptions) {
			this._subscriptionManager = new Web3SubscriptionManager(
				this.requestManager,
				registeredSubscriptions,
			);
		}

		if (accountProvider) {
			this._accountProvider = accountProvider;
		}

		if (wallet) {
			this._wallet = wallet;
		}
	}

	public get requestManager() {
		return this._requestManager;
	}

	public get subscriptionManager() {
		return this._subscriptionManager;
	}

	public get wallet() {
		return this._wallet;
	}

	public get accountProvider() {
		return this._accountProvider;
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	public static fromContextObject<T extends Web3Context<any>, T3 extends unknown[]>(
		this: Web3ContextConstructor<T, T3>,
		...args: [Web3ContextObject, ...T3]
	) {
		return new this(...(args.reverse() as [...T3, Web3ContextObject]));
	}

	public getContextObject(): Web3ContextObject<API, RegisteredSubs> {
		return {
			config: this.getConfig(),
			provider: this.provider,
			requestManager: this.requestManager,
			subscriptionManager: this.subscriptionManager,
			registeredSubscriptions: this.subscriptionManager
				?.registeredSubscriptions as RegisteredSubs,
			providers: this.providers,
			wallet: this.wallet,
			accountProvider: this.accountProvider,
		};
	}

	/**
	 * Use to create new object of any type extended by `Web3Context`
	 * and link it to current context. This can be used to initiate a global context object
	 * and then use it to create new objects of any type extended by `Web3Context`.
	 *
	 * @param ContextRef
	 * @param {...any} args
	 */
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	public use<T extends Web3Context<any>, T2 extends unknown[]>(
		ContextRef: Web3ContextConstructor<T, T2>,
		...args: [...T2]
	) {
		const newContextChild: T = new ContextRef(
			...([...args, this.getContextObject()] as unknown as [...T2, Web3ContextObject]),
		);

		this.on(Web3ConfigEvent.CONFIG_CHANGE, event => {
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
			newContextChild.setConfig({ [event.name]: event.newValue });
		});

		return newContextChild;
	}

	/**
	 * Link current context to another context.
	 *
	 * @param parentContext
	 */
	public link<T extends Web3Context<API, RegisteredSubs>>(parentContext: T) {
		this.setConfig(parentContext.getConfig());
		this._requestManager = parentContext.requestManager;
		this.provider = parentContext.provider;
		this._subscriptionManager = parentContext.subscriptionManager;
		this._wallet = parentContext.wallet;
		this._accountProvider = parentContext._accountProvider;

		parentContext.on(Web3ConfigEvent.CONFIG_CHANGE, event => {
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
			this.setConfig({ [event.name]: event.newValue });
		});
	}

	public get provider(): SupportedProviders<API> | string | undefined {
		return this.requestManager.provider;
	}

	public set provider(provider: SupportedProviders<API> | string | undefined) {
		this.requestManager.setProvider(provider);
	}

	public get currentProvider(): SupportedProviders<API> | string | undefined {
		return this.requestManager.provider;
	}

	public set currentProvider(provider: SupportedProviders<API> | string | undefined) {
		this.requestManager.setProvider(provider);
	}

	// eslint-disable-next-line class-methods-use-this
	public get givenProvider() {
		return Web3Context.givenProvider;
	}

	public setProvider(provider?: SupportedProviders<API> | string) {
		this.provider = provider;
	}

	public get BatchRequest(): new () => Web3BatchRequest {
		return Web3BatchRequest.bind(
			undefined,
			this._requestManager as unknown as Web3RequestManager,
		);
	}
}

// To avoid cycle dependency declare this type in this file
// TODO: When we have `web3-types` package we can share TransactionType
export type TransactionBuilder<
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	API extends Web3APISpec = any,
> = <ReturnType = Record<string, unknown>>(options: {
	transaction: Record<string, unknown>;
	web3Context: Web3Context<API>;
	privateKey?: HexString | Buffer;
}) => Promise<ReturnType>;
