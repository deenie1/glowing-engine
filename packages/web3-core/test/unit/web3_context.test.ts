// eslint-disable-next-line max-classes-per-file
import HttpProvider from 'web3-providers-http';
import { Web3Context } from '../../src/web3_context';
import { Web3RequestManager } from '../../src/web3_request_manager';

// eslint-disable-next-line @typescript-eslint/ban-types
class Context1 extends Web3Context<{}> {}
// eslint-disable-next-line @typescript-eslint/ban-types
class Context2 extends Web3Context<{}> {}

describe('Web3Context', () => {
	describe('constructor()', () => {
		it('should return providers on class level', () => {
			expect(Web3Context.providers).toBeDefined();
		});

		it('should return providers on instance level', () => {
			const context = new Web3Context('http://test.com');
			expect(context.providers).toBeDefined();
			expect(context.providers).toEqual(Web3Context.providers);
		});

		it('should create instance of request manager', () => {
			const context = new Web3Context('http://test.com');

			expect(context.requestManager).toBeInstanceOf(Web3RequestManager);
		});

		it('should return current provider from request manager', () => {
			const context = new Web3Context('http://test.com');

			expect(context.currentProvider).toBe(context.requestManager.provider);
		});

		it('should initialize the provider from options', () => {
			const context = new Web3Context({ provider: 'http://test.com' });

			expect(context.currentProvider).toBeInstanceOf(HttpProvider);
		});

		it('should set return current provider for the request manager', () => {
			const context = new Web3Context('http://test.com');

			context.currentProvider = 'http://test/abc';

			expect(context.currentProvider).toBeInstanceOf(HttpProvider);
		});
	});

	describe('getContextObject', () => {
		it('should return correct context object', () => {
			const context = new Context1('http://test/abc');

			expect(context.getContextObject()).toMatchSnapshot();
		});
	});

	describe('use', () => {
		it('should init child context with correct type', () => {
			const parent = new Context1('http://test/abc');
			const child = parent.use(Context2);

			expect(child).toBeInstanceOf(Context2);
		});

		it('should init context with correct configuration', () => {
			const parent = new Context1({
				provider: 'http://test/abc',
				config: { defaultNetworkId: 'my-network-id', defaultHardfork: 'my-fork' },
			});
			const child = parent.use(Context2);

			expect(child.defaultNetworkId).toBe('my-network-id');
			expect(child.defaultHardfork).toBe('my-fork');
		});

		it('should change config of child context', () => {
			const parent = new Context1('http://test/abc');
			const child = parent.use(Context2);

			parent.defaultNetworkId = 'my-network-id';

			expect(child.defaultNetworkId).toBe('my-network-id');
		});

		it('should use same instance of request manager', () => {
			const parent = new Context1('http://test/abc');
			const child = parent.use(Context2);

			expect(child.requestManager).toBe(parent.requestManager);
		});

		it('should use same instance of subscription manager', () => {
			const parent = new Context1({
				provider: 'http://test/abc',
				subscriptionManager: {} as any,
			});
			const child = parent.use(Context2);

			expect(child.subscriptionManager).toBe(parent.subscriptionManager);
		});

		it('should use same instance of provider', () => {
			const parent = new Context1('http://test/abc');
			const child = parent.use(Context2);

			expect(child.provider).toBe(parent.provider);
		});
	});

	describe('link', () => {
		it('should link context with correct configuration', () => {
			const parent = new Context1({
				provider: 'http://test/abc',
				config: { defaultNetworkId: 'my-network-id', defaultHardfork: 'my-fork' },
			});
			const child = new Context2('http://test/abc');

			child.link(parent);

			expect(child.defaultNetworkId).toBe('my-network-id');
			expect(child.defaultHardfork).toBe('my-fork');
		});

		it('should change config of child context', () => {
			const parent = new Context1('http://test/abc');
			const child = new Context2('http://test/abc');

			child.link(parent);

			parent.defaultNetworkId = 'my-network-id';

			expect(child.defaultNetworkId).toBe('my-network-id');
		});

		it('should use same instance of request manager', () => {
			const parent = new Context1('http://test/abc');
			const child = new Context2('http://test/abc');

			child.link(parent);

			expect(child.requestManager).toBe(parent.requestManager);
		});

		it('should use same instance of subscription manager', () => {
			const parent = new Context1({
				provider: 'http://test/abc',
				subscriptionManager: {} as any,
			});
			const child = new Context2('http://test/abc');

			child.link(parent);

			expect(child.subscriptionManager).toBe(parent.subscriptionManager);
		});

		it('should use same instance of provider', () => {
			const parent = new Context1('http://test/abc');
			const child = new Context2('http://test/abc');

			child.link(parent);

			expect(child.provider).toBe(parent.provider);
		});
	});
});
