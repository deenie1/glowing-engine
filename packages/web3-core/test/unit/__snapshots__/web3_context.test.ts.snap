// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web3Context getContextObject should return correct context object 1`] = `
Object {
  "accountProvider": undefined,
  "config": Object {
    "blockHeaderTimeout": 10,
    "defaultAccount": null,
    "defaultBlock": "latest",
    "defaultChain": "mainnet",
    "defaultCommon": null,
    "defaultHardfork": "london",
    "defaultMaxPriorityFeePerGas": "0x9502f900",
    "defaultNetworkId": null,
    "defaultTransactionType": "0x0",
    "handleRevert": false,
    "maxListenersWarningThreshold": 100,
    "transactionBlockTimeout": 50,
    "transactionBuilder": undefined,
    "transactionConfirmationBlocks": 24,
    "transactionConfirmationPollingInterval": null,
    "transactionPollingInterval": 1000,
    "transactionPollingTimeout": 750,
    "transactionReceiptPollingInterval": null,
    "transactionTypeParser": undefined,
  },
  "provider": HttpProvider {
    "clientUrl": "http://test/abc",
    "httpProviderOptions": undefined,
  },
  "providers": Object {
    "HttpProvider": [Function],
    "IpcProvider": [Function],
    "WebsocketProvider": [Function],
  },
  "registeredSubscriptions": undefined,
  "requestManager": Web3RequestManager {
    "_emitter": EventEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
      Symbol(kCapture): false,
    },
    "_provider": HttpProvider {
      "clientUrl": "http://test/abc",
      "httpProviderOptions": undefined,
    },
  },
  "subscriptionManager": undefined,
  "wallet": undefined,
}
`;
